File: AutoEmply/Services/StructureToLayoutConverter.cs
using AutoEmply.Models;
using System.Text.RegularExpressions;

namespace AutoEmply.Services;

/// <summary>
/// FormStructure(논리 구조) → LayoutSpec(픽셀 좌표) 변환기.
///
/// Phase 1에서 Claude가 비율 기반으로 추출한 논리 구조를,
/// Phase 2에서 Delphi DFM에 쓸 수 있는 픽셀 좌표 LayoutSpec으로 변환한다.
///
/// 변환 원리:
///   - 캔버스 너비 774px (좌우 여백 10px씩 제외)
///   - 열 너비 = 비율(fraction) × 캔버스 너비
///   - 행 높이 = HeightHint에 따른 고정 값 (standard=20, compact=14, tall=30)
///   - Z-Order: Rect(배경) → Line(테두리) → Text(라벨) 순으로 배치
/// </summary>
public sealed class StructureToLayoutConverter
{
    // ── 캔버스 상수 ──
    private const int CanvasLeft = 10;
    private const int ContentWidth = 774;

    // ── 크기 상수 ──
    private const int StandardRowHeight = 20;
    private const int CompactRowHeight = 14;
    private const int TallRowHeight = 30;
    private const int LabelHeight = 13;
    private const int TitleHeight = 28;
    private const int CellPaddingLeft = 4;
    private const int CellPaddingTop = 3;
    private const int SectionGap = 2;

    // ── 스타일 상수 ──
    private const string HeaderBgColor = "#D8E4F0";
    private const string BorderColor = "#000000";

    private static readonly Regex InternalNameRegex = new("^(Q(?:lb|sp|sh|img)\\d*_[A-Za-z0-9_]+)$", RegexOptions.Compiled | RegexOptions.IgnoreCase);
    private static readonly Regex FieldCodeRegex = new("([A-Za-z]{2,4}\\d{2,4})", RegexOptions.Compiled);

    // ═══════════════════════════════════════════
    //  공개 API
    // ═══════════════════════════════════════════

    public LayoutSpec Convert(FormStructure structure)
    {
        var items = new List<LayoutItem>();
        var currentY = 6;

        // 제목
        if (!string.IsNullOrWhiteSpace(structure.Title))
        {
            items.Add(MakeLabel(structure.Title,
                CanvasLeft + ContentWidth / 4, currentY, ContentWidth / 2, TitleHeight,
                structure.TitleFontSize, bold: true, align: "Center", name: "Qlb3_TITLE"));
            currentY += TitleHeight + 6;
        }

        // 섹션들
        foreach (var section in structure.Sections)
        {
            currentY = RenderSection(section, items, currentY);
            currentY += SectionGap;
        }

        // 푸터
        if (structure.Footer is { Count: > 0 })
        {
            currentY += 4;
            RenderFooter(structure.Footer, items, currentY);
        }

        return new LayoutSpec { Items = items };
    }

    // ═══════════════════════════════════════════
    //  섹션 렌더링
    // ═══════════════════════════════════════════

    private int RenderSection(FormSection section, List<LayoutItem> items, int startY) =>
        (section.SectionType ?? "table").ToLowerInvariant() switch
        {
            "table" => RenderTableSection(section, items, startY),
            _ => RenderFreeformSection(section, items, startY)
        };

    private int RenderTableSection(FormSection section, List<LayoutItem> items, int startY)
    {
        var table = section.Table;
        if (table is null || table.Columns.Count == 0 || table.Rows.Count == 0)
            return startY;

        // 테이블 경계 계산
        var (tableLeft, tableWidth) = table.FullWidth
            ? (CanvasLeft, ContentWidth)
            : (CanvasLeft + (int)Math.Round(table.LeftFraction * ContentWidth),
               (int)Math.Round(table.WidthFraction * ContentWidth));

        var colBounds = CalculateColumnBoundaries(table.Columns, tableLeft, tableWidth);
        var rowYs = CalculateRowYPositions(table.Rows, startY);
        var totalHeight = rowYs[^1] - startY;

        // 1. 배경 사각형 (Z-Order 최하위)
        RenderBackgrounds(table, items, colBounds, rowYs, tableLeft, tableWidth, startY, totalHeight);

        // 2. 테두리 선
        RenderBorders(table, items, colBounds, rowYs, tableLeft, tableWidth, startY, totalHeight);

        // 3. 셀 텍스트
        RenderCellTexts(table, items, colBounds, rowYs);

        return rowYs[^1];
    }

    // ── 테이블: 배경 ──

    private static void RenderBackgrounds(
        TableDef table, List<LayoutItem> items, int[] colBounds, List<int> rowYs,
        int tableLeft, int tableWidth, int startY, int totalHeight)
    {
        // 헤더 열 배경 (전체 높이)
        for (var c = 0; c < table.Columns.Count; c++)
        {
            if (table.Columns[c].IsHeaderColumn)
                items.Add(MakeRect(colBounds[c] + 1, startY + 1,
                    colBounds[c + 1] - colBounds[c] - 1, totalHeight - 1, HeaderBgColor));
        }

        // 헤더 행 / 개별 셀 배경
        for (var r = 0; r < table.Rows.Count; r++)
        {
            var rowHeight = rowYs[r + 1] - rowYs[r];
            if (table.Rows[r].IsHeaderRow)
            {
                items.Add(MakeRect(tableLeft + 1, rowYs[r] + 1, tableWidth - 1, rowHeight - 1, HeaderBgColor));
                continue;
            }

            var colIndex = 0;
            foreach (var cell in table.Rows[r].Cells)
            {
                if (cell.HasBackground && colIndex < colBounds.Length - 1)
                {
                    var spanEnd = Math.Min(colIndex + Math.Max(1, cell.ColSpan), colBounds.Length - 1);
                    items.Add(MakeRect(colBounds[colIndex] + 1, rowYs[r] + 1,
                        colBounds[spanEnd] - colBounds[colIndex] - 1, rowHeight - 1, HeaderBgColor));
                }
                colIndex += Math.Max(1, cell.ColSpan);
            }
        }
    }

    // ── 테이블: 테두리 ──

    private static void RenderBorders(
        TableDef table, List<LayoutItem> items, int[] colBounds, List<int> rowYs,
        int tableLeft, int tableWidth, int startY, int totalHeight)
    {
        // 가로선: 상단, 하단, 행 사이
        for (var r = 0; r <= table.Rows.Count; r++)
            items.Add(MakeHLine(tableLeft, rowYs[r], tableWidth));

        // 세로선: 좌측, 우측, 열 사이
        var tableRight = tableLeft + tableWidth;
        for (var c = 0; c <= table.Columns.Count; c++)
        {
            var x = c < colBounds.Length ? colBounds[c] : tableRight;
            items.Add(MakeVLine(x, startY, totalHeight));
        }
    }

    // ── 테이블: 셀 텍스트 ──

    private static void RenderCellTexts(
        TableDef table, List<LayoutItem> items, int[] colBounds, List<int> rowYs)
    {
        for (var r = 0; r < table.Rows.Count; r++)
        {
            var colIndex = 0;
            foreach (var cell in table.Rows[r].Cells)
            {
                if (colIndex >= colBounds.Length - 1) break;

                var colSpan = Math.Max(1, cell.ColSpan);
                var spanEnd = Math.Min(colIndex + colSpan, colBounds.Length - 1);
                var cellLeft = colBounds[colIndex] + CellPaddingLeft;
                var cellTop = rowYs[r] + CellPaddingTop;
                var cellWidth = colBounds[spanEnd] - colBounds[colIndex] - CellPaddingLeft * 2;

                var displayText = ResolveDisplayText(cell.Text, cell.FieldName);
                var componentName = ResolveComponentName(displayText, cell.FieldName);

                if (!string.IsNullOrWhiteSpace(displayText))
                {
                    items.Add(MakeLabel(displayText, cellLeft, cellTop,
                        Math.Max(1, cellWidth), LabelHeight,
                        cell.FontSize > 0 ? cell.FontSize : 9,
                        cell.Bold, cell.Align, cell.TextColor, componentName));
                }

                colIndex += colSpan;
            }
        }
    }

    // ═══════════════════════════════════════════
    //  자유 배치 / 푸터 렌더링
    // ═══════════════════════════════════════════

    private int RenderFreeformSection(FormSection section, List<LayoutItem> items, int startY)
    {
        if (section.Elements is null or { Count: 0 }) return startY;

        var maxBottom = section.Elements.Max(e => e.YFraction + e.HeightFraction);
        var sectionHeight = Math.Max(StandardRowHeight, (int)Math.Round(maxBottom * 200));

        foreach (var el in section.Elements)
        {
            var left = CanvasLeft + (int)Math.Round(el.XFraction * ContentWidth);
            var top = startY + (int)Math.Round(el.YFraction * sectionHeight);
            var width = Math.Max(1, (int)Math.Round(el.WidthFraction * ContentWidth));
            var height = Math.Max(LabelHeight, (int)Math.Round(el.HeightFraction * sectionHeight));

            switch ((el.ElementType ?? "text").ToLowerInvariant())
            {
                case "line":
                    items.Add(MakeHLine(left, top, width));
                    break;
                case "image":
                    items.Add(new LayoutItem { Type = "Image", Left = left, Top = top, Width = width, Height = height, Stretch = true });
                    break;
                default:
                    items.Add(MakeLabel(el.Text ?? "", left, top, width, LabelHeight,
                        el.FontSize > 0 ? el.FontSize : 9, el.Bold, el.Align));
                    break;
            }
        }

        return startY + sectionHeight;
    }

    private static int RenderFooter(List<FooterElement> footerElements, List<LayoutItem> items, int startY)
    {
        var y = startY;
        foreach (var el in footerElements)
        {
            var left = CanvasLeft + (int)Math.Round(el.XFraction * ContentWidth);
            var width = Math.Max(1, (int)Math.Round(el.WidthFraction * ContentWidth));

            switch ((el.ElementType ?? "text").ToLowerInvariant())
            {
                case "image":
                    items.Add(new LayoutItem { Type = "Image", Left = left, Top = y, Width = width, Height = 30, Stretch = true });
                    y += 34;
                    break;
                case "signature_line":
                    items.Add(MakeHLine(left, y + 10, width));
                    y += 14;
                    break;
                default:
                    items.Add(MakeLabel(el.Text ?? "", left, y, width, LabelHeight,
                        el.FontSize > 0 ? el.FontSize : 9, el.Bold, el.Align));
                    y += LabelHeight + 2;
                    break;
            }
        }
        return y;
    }

    // ═══════════════════════════════════════════
    //  계산 헬퍼
    // ═══════════════════════════════════════════

    private static int[] CalculateColumnBoundaries(List<ColumnDef> columns, int tableLeft, int tableWidth)
    {
        var totalFraction = columns.Sum(c => c.WidthFraction);
        var normFactor = totalFraction > 0 ? 1.0 / totalFraction : 1.0;

        var bounds = new int[columns.Count + 1];
        bounds[0] = tableLeft;

        var accum = 0.0;
        for (var i = 0; i < columns.Count; i++)
        {
            accum += columns[i].WidthFraction * normFactor;
            bounds[i + 1] = tableLeft + (int)Math.Round(accum * tableWidth);
        }
        bounds[^1] = tableLeft + tableWidth; // 마지막 열은 정확하게

        return bounds;
    }

    private static List<int> CalculateRowYPositions(List<TableRow> rows, int startY)
    {
        var positions = new List<int> { startY };
        foreach (var row in rows)
            positions.Add(positions[^1] + ResolveRowHeight(row));
        return positions;
    }

    private static int ResolveRowHeight(TableRow row) =>
        (row.HeightHint ?? "standard").Trim().ToLowerInvariant() switch
        {
            "standard" => StandardRowHeight,
            "compact" => CompactRowHeight,
            "tall" => TallRowHeight,
            _ when int.TryParse(row.HeightHint, out var h) => Math.Clamp(h, 10, 300),
            _ => StandardRowHeight
        };

    // ═══════════════════════════════════════════
    //  텍스트 / 컴포넌트 이름 결정
    // ═══════════════════════════════════════════

    /// <summary>셀의 표시 텍스트를 결정. 내부 컴포넌트 식별자는 캡션으로 노출하지 않는다.</summary>
    private static string ResolveDisplayText(string? text, string? fieldName)
    {
        var visible = (text ?? "").Trim();
        if (!string.IsNullOrWhiteSpace(visible)) return visible;

        var field = (fieldName ?? "").Trim();
        if (string.IsNullOrWhiteSpace(field)) return string.Empty;

        return InternalNameRegex.IsMatch(field) ? string.Empty : field;
    }

    /// <summary>Delphi 컴포넌트 이름을 fieldName이나 표시 텍스트에서 유추.</summary>
    private static string? ResolveComponentName(string? displayText, string? fieldName)
    {
        var field = (fieldName ?? "").Trim();
        if (!string.IsNullOrWhiteSpace(field))
        {
            var match = FieldCodeRegex.Match(field);
            if (match.Success)
                return $"Qlb3_{match.Groups[1].Value.ToUpperInvariant()}";
        }

        var visible = (displayText ?? "").Trim();
        if (!string.IsNullOrWhiteSpace(visible))
        {
            var sanitized = SanitizeIdentifier(visible);
            if (!string.IsNullOrWhiteSpace(sanitized))
                return $"Qlb3_{sanitized}";
        }

        return null;
    }

    private static string SanitizeIdentifier(string raw)
    {
        var chars = raw
            .Select(ch => (ch is >= 'A' and <= 'Z') || (ch is >= 'a' and <= 'z') || (ch is >= '0' and <= '9') ? ch : '_')
            .ToArray();
        var candidate = new string(chars).Trim('_');
        if (string.IsNullOrWhiteSpace(candidate)) return string.Empty;

        while (candidate.Contains("__", StringComparison.Ordinal))
            candidate = candidate.Replace("__", "_", StringComparison.Ordinal);
        if (!char.IsLetter(candidate[0]) && candidate[0] != '_')
            candidate = $"N_{candidate}";

        return candidate.Length > 40 ? candidate[..40] : candidate;
    }

    // ═══════════════════════════════════════════
    //  LayoutItem 팩토리
    // ═══════════════════════════════════════════

    private static LayoutItem MakeLabel(string caption, int left, int top, int width, int height,
        int fontSize = 9, bool bold = false, string align = "Left", string? textColor = null, string? name = null) => new()
    {
        Name = name, Type = "Text",
        Left = Math.Max(0, left), Top = Math.Max(0, top),
        Width = Math.Max(1, width), Height = Math.Max(1, height),
        Caption = caption, FontSize = Math.Clamp(fontSize, 6, 24),
        Bold = bold, Align = align, Transparent = true,
        TextColor = textColor ?? "#000000"
    };

    private static LayoutItem MakeHLine(int left, int top, int width) => new()
    {
        Type = "Line", Left = Math.Max(0, left), Top = Math.Max(0, top),
        Width = Math.Max(1, width), Height = 1,
        Orientation = "H", Thickness = 1, StrokeColor = BorderColor
    };

    private static LayoutItem MakeVLine(int left, int top, int height) => new()
    {
        Type = "Line", Left = Math.Max(0, left), Top = Math.Max(0, top),
        Width = 1, Height = Math.Max(1, height),
        Orientation = "V", Thickness = 1, StrokeColor = BorderColor
    };

    private static LayoutItem MakeRect(int left, int top, int width, int height, string fillColor) => new()
    {
        Type = "Rect", Left = Math.Max(0, left), Top = Math.Max(0, top),
        Width = Math.Max(1, width), Height = Math.Max(1, height),
        FillColor = fillColor, Filled = true, StrokeColor = fillColor, Thickness = 1
    };
}

File: AutoEmply-server/src/main/java/health/autoemplyserver/service/StructureToLayoutConverter.java
package health.autoemplyserver.service;

import health.autoemplyserver.model.ColumnDef;
import health.autoemplyserver.model.FooterElement;
import health.autoemplyserver.model.FormSection;
import health.autoemplyserver.model.FormStructure;
import health.autoemplyserver.model.FreeformElement;
import health.autoemplyserver.model.LayoutItem;
import health.autoemplyserver.model.LayoutSpec;
import health.autoemplyserver.model.TableCell;
import health.autoemplyserver.model.TableDef;
import health.autoemplyserver.model.TableRow;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.springframework.stereotype.Component;

@Component
public class StructureToLayoutConverter {

    private static final int CANVAS_LEFT = 10;
    private static final int CONTENT_WIDTH = 774;
    private static final int STANDARD_ROW_HEIGHT = 20;
    private static final int COMPACT_ROW_HEIGHT = 14;
    private static final int TALL_ROW_HEIGHT = 30;
    private static final int LABEL_HEIGHT = 13;
    private static final int TITLE_HEIGHT = 28;
    private static final int CELL_PADDING_LEFT = 4;
    private static final int CELL_PADDING_TOP = 3;
    private static final int SECTION_GAP = 2;
    private static final String HEADER_BG_COLOR = "#D8E4F0";
    private static final String BORDER_COLOR = "#000000";
    private static final Pattern FIELD_CODE_REGEX = Pattern.compile("([A-Za-z]{2,4}\\d{2,4})");

    public LayoutSpec convert(FormStructure structure) {
        List<LayoutItem> items = new ArrayList<>();
        int currentY = 6;

        if (structure.getTitle() != null && !structure.getTitle().isBlank()) {
            items.add(makeLabel(structure.getTitle(), CANVAS_LEFT + CONTENT_WIDTH / 4, currentY, CONTENT_WIDTH / 2, TITLE_HEIGHT,
                structure.getTitleFontSize(), true, "Center", "#000000", "Qlb3_TITLE"));
            currentY += TITLE_HEIGHT + 6;
        }

        for (FormSection section : structure.getSections()) {
            currentY = renderSection(section, items, currentY);
            currentY += SECTION_GAP;
        }

        if (structure.getFooter() != null && !structure.getFooter().isEmpty()) {
            currentY += 4;
            renderFooter(structure.getFooter(), items, currentY);
        }

        LayoutSpec layoutSpec = new LayoutSpec();
        layoutSpec.setItems(items);
        return layoutSpec;
    }

    private int renderSection(FormSection section, List<LayoutItem> items, int startY) {
        String sectionType = section.getSectionType() == null ? "table" : section.getSectionType().toLowerCase();
        if ("table".equals(sectionType)) {
            return renderTableSection(section, items, startY);
        }
        return renderFreeformSection(section, items, startY);
    }

    private int renderTableSection(FormSection section, List<LayoutItem> items, int startY) {
        TableDef table = section.getTable();
        if (table == null || table.getColumns().isEmpty() || table.getRows().isEmpty()) {
            return startY;
        }

        int tableLeft = table.isFullWidth() ? CANVAS_LEFT : CANVAS_LEFT + (int) Math.round(table.getLeftFraction() * CONTENT_WIDTH);
        int tableWidth = table.isFullWidth() ? CONTENT_WIDTH : (int) Math.round(table.getWidthFraction() * CONTENT_WIDTH);
        int[] colBounds = calculateColumnBoundaries(table.getColumns(), tableLeft, tableWidth);
        List<Integer> rowYs = calculateRowPositions(table.getRows(), startY);
        int totalHeight = rowYs.getLast() - startY;

        renderBackgrounds(table, items, colBounds, rowYs, tableLeft, tableWidth, startY, totalHeight);
        renderBorders(table, items, colBounds, rowYs, tableLeft, tableWidth, startY, totalHeight);
        renderCellTexts(table, items, colBounds, rowYs);

        return rowYs.getLast();
    }

    private void renderBackgrounds(TableDef table, List<LayoutItem> items, int[] colBounds, List<Integer> rowYs, int tableLeft, int tableWidth, int startY, int totalHeight) {
        for (int col = 0; col < table.getColumns().size(); col++) {
            if (table.getColumns().get(col).isHeaderColumn()) {
                items.add(makeRect(colBounds[col] + 1, startY + 1, colBounds[col + 1] - colBounds[col] - 1, totalHeight - 1, HEADER_BG_COLOR));
            }
        }

        for (int rowIndex = 0; rowIndex < table.getRows().size(); rowIndex++) {
            TableRow row = table.getRows().get(rowIndex);
            int rowHeight = rowYs.get(rowIndex + 1) - rowYs.get(rowIndex);
            if (row.isHeaderRow()) {
                items.add(makeRect(tableLeft + 1, rowYs.get(rowIndex) + 1, tableWidth - 1, rowHeight - 1, HEADER_BG_COLOR));
                continue;
            }
            int colIndex = 0;
            for (TableCell cell : row.getCells()) {
                if (cell.isHasBackground() && colIndex < colBounds.length - 1) {
                    int spanEnd = Math.min(colIndex + Math.max(1, cell.getColSpan()), colBounds.length - 1);
                    items.add(makeRect(colBounds[colIndex] + 1, rowYs.get(rowIndex) + 1,
                        colBounds[spanEnd] - colBounds[colIndex] - 1, rowHeight - 1, HEADER_BG_COLOR));
                }
                colIndex += Math.max(1, cell.getColSpan());
            }
        }
    }

    private void renderBorders(TableDef table, List<LayoutItem> items, int[] colBounds, List<Integer> rowYs, int tableLeft, int tableWidth, int startY, int totalHeight) {
        for (int row = 0; row <= table.getRows().size(); row++) {
            items.add(makeHLine(tableLeft, rowYs.get(row), tableWidth));
        }
        for (int col = 0; col <= table.getColumns().size(); col++) {
            int x = col < colBounds.length ? colBounds[col] : tableLeft + tableWidth;
            items.add(makeVLine(x, startY, totalHeight));
        }
    }

    private void renderCellTexts(TableDef table, List<LayoutItem> items, int[] colBounds, List<Integer> rowYs) {
        for (int rowIndex = 0; rowIndex < table.getRows().size(); rowIndex++) {
            int colIndex = 0;
            for (TableCell cell : table.getRows().get(rowIndex).getCells()) {
                if (colIndex >= colBounds.length - 1) {
                    break;
                }
                int spanEnd = Math.min(colIndex + Math.max(1, cell.getColSpan()), colBounds.length - 1);
                int cellLeft = colBounds[colIndex] + CELL_PADDING_LEFT;
                int cellTop = rowYs.get(rowIndex) + CELL_PADDING_TOP;
                int cellWidth = colBounds[spanEnd] - colBounds[colIndex] - CELL_PADDING_LEFT * 2;
                String displayText = resolveDisplayText(cell.getText(), cell.getFieldName());
                if (displayText != null && !displayText.isBlank()) {
                    items.add(makeLabel(displayText, cellLeft, cellTop, Math.max(1, cellWidth), LABEL_HEIGHT,
                        cell.getFontSize() > 0 ? cell.getFontSize() : 9, cell.isBold(), cell.getAlign(), cell.getTextColor(), resolveComponentName(displayText, cell.getFieldName())));
                }
                colIndex += Math.max(1, cell.getColSpan());
            }
        }
    }

    private int renderFreeformSection(FormSection section, List<LayoutItem> items, int startY) {
        if (section.getElements() == null || section.getElements().isEmpty()) {
            return startY;
        }

        double maxBottom = section.getElements().stream().mapToDouble(element -> element.getYFraction() + element.getHeightFraction()).max().orElse(0.2d);
        int sectionHeight = Math.max(STANDARD_ROW_HEIGHT, (int) Math.round(maxBottom * 200));

        for (FreeformElement element : section.getElements()) {
            int left = CANVAS_LEFT + (int) Math.round(element.getXFraction() * CONTENT_WIDTH);
            int top = startY + (int) Math.round(element.getYFraction() * sectionHeight);
            int width = Math.max(1, (int) Math.round(element.getWidthFraction() * CONTENT_WIDTH));
            int height = Math.max(LABEL_HEIGHT, (int) Math.round(element.getHeightFraction() * sectionHeight));

            switch ((element.getElementType() == null ? "text" : element.getElementType()).toLowerCase()) {
                case "line" -> items.add(makeHLine(left, top, width));
                case "image" -> {
                    LayoutItem item = new LayoutItem();
                    item.setType("Image");
                    item.setLeft(left);
                    item.setTop(top);
                    item.setWidth(width);
                    item.setHeight(height);
                    item.setStretch(true);
                    items.add(item);
                }
                default -> items.add(makeLabel(element.getText() == null ? "" : element.getText(), left, top, width, LABEL_HEIGHT,
                    element.getFontSize(), element.isBold(), element.getAlign(), "#000000", null));
            }
        }
        return startY + sectionHeight;
    }

    private void renderFooter(List<FooterElement> footerElements, List<LayoutItem> items, int startY) {
        int y = startY;
        for (FooterElement element : footerElements) {
            int left = CANVAS_LEFT + (int) Math.round(element.getXFraction() * CONTENT_WIDTH);
            int width = Math.max(1, (int) Math.round(element.getWidthFraction() * CONTENT_WIDTH));
            switch ((element.getElementType() == null ? "text" : element.getElementType()).toLowerCase()) {
                case "image" -> {
                    LayoutItem image = new LayoutItem();
                    image.setType("Image");
                    image.setLeft(left);
                    image.setTop(y);
                    image.setWidth(width);
                    image.setHeight(30);
                    image.setStretch(true);
                    items.add(image);
                    y += 34;
                }
                case "signature_line" -> {
                    items.add(makeHLine(left, y + 10, width));
                    y += 14;
                }
                default -> {
                    items.add(makeLabel(element.getText() == null ? "" : element.getText(), left, y, width, LABEL_HEIGHT, element.getFontSize(), element.isBold(), element.getAlign(), "#000000", null));
                    y += LABEL_HEIGHT + 2;
                }
            }
        }
    }

    private int[] calculateColumnBoundaries(List<ColumnDef> columns, int tableLeft, int tableWidth) {
        int[] bounds = new int[columns.size() + 1];
        bounds[0] = tableLeft;
        double total = columns.stream().mapToDouble(ColumnDef::getWidthFraction).sum();
        double factor = total > 0 ? 1.0d / total : 1.0d;
        double accum = 0;
        for (int index = 0; index < columns.size(); index++) {
            accum += columns.get(index).getWidthFraction() * factor;
            bounds[index + 1] = tableLeft + (int) Math.round(accum * tableWidth);
        }
        bounds[columns.size()] = tableLeft + tableWidth;
        return bounds;
    }

    private List<Integer> calculateRowPositions(List<TableRow> rows, int startY) {
        List<Integer> positions = new ArrayList<>();
        positions.add(startY);
        for (TableRow row : rows) {
            positions.add(positions.getLast() + resolveRowHeight(row));
        }
        return positions;
    }

    private int resolveRowHeight(TableRow row) {
        String hint = row.getHeightHint() == null ? "standard" : row.getHeightHint().trim().toLowerCase();
        return switch (hint) {
            case "compact" -> COMPACT_ROW_HEIGHT;
            case "tall" -> TALL_ROW_HEIGHT;
            case "standard" -> STANDARD_ROW_HEIGHT;
            default -> {
                try {
                    yield Math.max(10, Math.min(300, Integer.parseInt(hint)));
                } catch (NumberFormatException exception) {
                    yield STANDARD_ROW_HEIGHT;
                }
            }
        };
    }

    private String resolveDisplayText(String text, String fieldName) {
        if (text != null && !text.trim().isBlank()) {
            return text.trim();
        }
        return fieldName == null ? "" : fieldName.trim();
    }

    private String resolveComponentName(String displayText, String fieldName) {
        if (fieldName != null && !fieldName.isBlank()) {
            Matcher matcher = FIELD_CODE_REGEX.matcher(fieldName.trim());
            if (matcher.find()) {
                return "Qlb3_" + matcher.group(1).toUpperCase();
            }
        }
        if (displayText == null || displayText.isBlank()) {
            return null;
        }
        String candidate = displayText.replaceAll("[^A-Za-z0-9]", "_").replaceAll("_+", "_").replaceAll("^_+|_+$", "");
        return candidate.isBlank() ? null : "Qlb3_" + candidate.substring(0, Math.min(40, candidate.length()));
    }

    private LayoutItem makeLabel(String caption, int left, int top, int width, int height, int fontSize, boolean bold, String align, String textColor, String name) {
        LayoutItem item = new LayoutItem();
        item.setName(name);
        item.setType("Text");
        item.setLeft(Math.max(0, left));
        item.setTop(Math.max(0, top));
        item.setWidth(Math.max(1, width));
        item.setHeight(Math.max(1, height));
        item.setCaption(caption);
        item.setFontSize(Math.max(6, Math.min(24, fontSize)));
        item.setBold(bold);
        item.setAlign(align == null ? "Left" : align);
        item.setTransparent(true);
        item.setTextColor(textColor);
        return item;
    }

    private LayoutItem makeHLine(int left, int top, int width) {
        LayoutItem item = new LayoutItem();
        item.setType("Line");
        item.setLeft(Math.max(0, left));
        item.setTop(Math.max(0, top));
        item.setWidth(Math.max(1, width));
        item.setHeight(1);
        item.setOrientation("H");
        item.setThickness(1);
        item.setStrokeColor(BORDER_COLOR);
        return item;
    }

    private LayoutItem makeVLine(int left, int top, int height) {
        LayoutItem item = new LayoutItem();
        item.setType("Line");
        item.setLeft(Math.max(0, left));
        item.setTop(Math.max(0, top));
        item.setWidth(1);
        item.setHeight(Math.max(1, height));
        item.setOrientation("V");
        item.setThickness(1);
        item.setStrokeColor(BORDER_COLOR);
        return item;
    }

    private LayoutItem makeRect(int left, int top, int width, int height, String fillColor) {
        LayoutItem item = new LayoutItem();
        item.setType("Rect");
        item.setLeft(Math.max(0, left));
        item.setTop(Math.max(0, top));
        item.setWidth(Math.max(1, width));
        item.setHeight(Math.max(1, height));
        item.setFillColor(fillColor);
        item.setFilled(true);
        item.setStrokeColor(fillColor);
        item.setThickness(1);
        return item;
    }
}
